mantidproject/mantid
# LoadEventNexus with compression refactor

This is a heavy rework of how `LoadEventNexus` compresses data. Previously `LoadEventNexus` would create all the events with no time then internally call the methods on `EventList` to compress the events. This accumulates the weight of the events then creates the weighted events directly. The load takes longer, but it uses less memory and workflows as a whole are faster.

The other thing to note (and added to the algorithm docs) is that `LoadEventNexus` with compression creates workspaces that are different from `LoadEventNexus` followed by `CompressEvents`. This is because the comparison of time-of-flight within tolerance is slightly different. Since the tolerance is significantly smaller than the bin size this should not have a practical effect.

Refs #36594 and [EWM3412](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=3412)

#### Benchmark results

As stated above, this slows down `LoadEventNexus`, but speeds up workflows. This section details workflows evaluated.

##### VULCAN
VULCAN_218092.nxs.h5 (10GB)

```python
from mantid.simpleapi import *

direc='/path/to/data/'
nxs_file=direc+'VULCAN_218092.nxs.h5' # 10G
cal_file=direc+'B123456DIFCs-12Cross-3456Cal_v4.h5'

LoadEventNexus(Filename=nxs_file, OutputWorkspace='wksp', NumberOfBins=1, CompressTolerance=-0.0002,
               BlockList="Phase\\*,Speed\\*,BL\\*:Chop:\\*,chopper\\*TDC")
LoadDiffCal(InputWorkspace='wksp', Filename=cal_file,
            #MakeCalWorkspace=False,
            WorkspaceName='VULCAN')
AlignAndFocusPowder(InputWorkspace='wksp', OutputWorkspace='wksp',
                             GroupingWorkspace="VULCAN_group",
                             CalibrationWorkspace='VULCAN_cal',
                             MaskWorkspace='VULCAN_mask',
                             CompressTolerance=0.1,
                             Dspacing=True,
                             Params=(.1,-.001,5)) # preserving events so can rebin later
```
vulcanperf.py Load w/ compression peak memory 14.1GB
load 20.3+-0.6s full workflow 30.0+-0.7s

```python
from mantid.simpleapi import *

direc='/path/to/data/'
nxs_file=direc+'VULCAN_218092.nxs.h5' # 10G
cal_file=direc+'B123456DIFCs-12Cross-3456Cal_v4.h5'

LoadEventNexus(Filename=nxs_file, OutputWorkspace='wksp', NumberOfBins=1,
               BlockList="Phase\\*,Speed\\*,BL\\*:Chop:\\*,chopper\\*TDC")
CompressEvents(InputWorkspace='wksp', OutputWorkspace='wksp', Tolerance=.1)
LoadDiffCal(InputWorkspace='wksp', Filename=cal_file,
            #MakeCalWorkspace=False,
            WorkspaceName='VULCAN')
AlignAndFocusPowder(InputWorkspace='wksp', OutputWorkspace='wksp',
                             GroupingWorkspace="VULCAN_group",
                             CalibrationWorkspace='VULCAN_cal',
                             MaskWorkspace='VULCAN_mask',
                             CompressTolerance=0.1,
                             Dspacing=True,
                             Params=(.1,-.001,5)) # preserving events so can rebin later
```
vulcanperf.py Load then compression peak memory 27.7GB released at the end of load down to 20GB
load 15.6+-0.6s full workflow 35.6+-1.0s

##### SNAP
* SNAP_57514.nxs.h5 (7.7GB) is called full
* SNAP_57514.lite.nxs.h5 (7.7GB) is called lite

```python
# Load with compression
direc='/place/with/data/'
nxs_file=direc+'SNAP_57514.lite.nxs.h5'
nxs_file=direc+'SNAP_57514.nxs.h5'

LoadEventNexus(Filename=nxs_file, OutputWorkspace='wksp', NumberOfBins=1, CompressTolerance=-0.0002)
CompressEvents(InputWorkspace='wksp', OutputWorkspace='wksp', Tolerance=0.1)
CreateGroupingWorkspace(InputWorkspace="wksp", OutputWorkspace="SNAP_group", GroupDetectorsBy="bank")
AlignAndFocusPowder(InputWorkspace='wksp', OutputWorkspace='wksp',
                             GroupingWorkspace="SNAP_group",
                             CompressTolerance=0.1,
                             Dspacing=True,
                             Params=(.1,-.0001,5)) # preserving events so can rebin later
```
snapperf.py Load w/ compression memory from load 13.0GB peak in convertunits 23GB
load 15.5+-0.3s full workflow 25.8+-0.5s

snapperf.py Load w/ compression memory peak memory 3.4GB
load 11.9+-0.2s full workflow 14.5+-0.2s

```python
# Load then with compress
direc='/place/with/data/'
nxs_file=direc+'SNAP_57514.lite.nxs.h5'
nxs_file=direc+'SNAP_57514.nxs.h5'

LoadEventNexus(Filename=nxs_file, OutputWorkspace='wksp', NumberOfBins=1)
CompressEvents(InputWorkspace='wksp', OutputWorkspace='wksp', Tolerance=0.1)
CreateGroupingWorkspace(InputWorkspace="wksp", OutputWorkspace="SNAP_group", GroupDetectorsBy="bank")
AlignAndFocusPowder(InputWorkspace='wksp', OutputWorkspace='wksp',
                             GroupingWorkspace="SNAP_group",
                             CompressTolerance=0.1,
                             Dspacing=True,
                             Params=(.1,-.0001,5)) # preserving events so can rebin later
```
snapperf.py Load then compression 13.6GB in load peak in convertunits 26GB
load 11.5+-0.1s full workflow 23.9+-0.2s

snapperf.py Load then compression peak memory 11.6GB
load 7.0+-0.0s full workflow 17.6+-0.1s

not compressing before `AlignAndFocusPowder` is slower.

### To test:

Try workflows that currently use LoadEventNexus w/ compression and see that they continue to work.



---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
